back/src/adapters/secondary/offer/InMemoryEstablishmentLeadRepository.ts

```diff
@@ -21,7 +26,39 @@ export class InMemoryEstablishmentLeadRepository
   public async getBySiret(
     siret: SiretDto,
   ): Promise<EstablishmentLead | undefined> {
-    return this.#establishmentLeads[siret];
+    const establishmentLead = this.#establishmentLeads[siret];
+
+    if (!establishmentLead) return;
+
+    const sortedEvents = sortBy(prop("occurredAt"))(establishmentLead?.events);
+    const lastEventKind = sortedEvents[sortedEvents.length - 1].kind;
+    return { ...establishmentLead, lastEventKind };
+  }
+
+  public getSiretsByUniqLastEventKind({
+    kind,
+    beforeDate,
+  }: EstablishmentLeadReminderParams): Promise<SiretDto[]> {
+    const sirets = Object.values(this.#establishmentLeads)
+      .filter(
+        ({ lastEventKind, events }) =>
+          lastEventKind === kind &&
+          events.filter((event) => event.kind === kind).length === 1,
+      )
+      .filter(({ events }) => {
+        return beforeDate
+          ? events.filter(
+              (event) => event.kind === kind && event.occurredAt <= beforeDate,
+            ).length > 0
+          : true;
+      })
+      .map(({ siret, events }) => {
+        console.log("beforeDate", beforeDate);
+        console.log("events", events);
+        return siret;
+      });
+    console.log("sirets", sirets);
+    return Promise.resolve(sirets);
```

Comment:
Pourquoi cette histoire de longueur === 1 ? 

On veut juste le dernier qui match non ? Et on doit pouvoir le faire en un seul filtre.

Et on peut faire une fonction async pour éviter le promise.resolve.

Un truc du genre :

```suggestion
    return Object.values(this.#establishmentLeads)
      .filter(
        ({ lastEventKind, events }) => {
          const lastEvent = events.at(-1)
          if(lastEvent?.kind !== kind) return false;
          if(!beforeDate) return true;
          return lastEvent.occurredAt <= beforeDate
      })
      .map(({ siret }) => siret);
```

Reply:
> Pourquoi cette histoire de longueur === 1 ? 
> On veut juste le dernier qui match non ? Et on doit pouvoir le faire en un seul filtre.

On vérifie qu'il n'y a qu'une seul event du type `kind` car dans le cas du `reminder-sent` il peut y en avoir deux une fois les deux relances faite

Reply:
Pour reprendre ta suggestion, il faudrait rajouter un tri sur les events + vérifier que l'event n'ait eu lieu qu'une seule fois:

```js
return Object.values(this.#establishmentLeads)
      .filter(
        ({ events }) => {
          const sortedEvents = events.sort((event1, event2) => event2.occurredAt.getMilliseconds() - event1.occurredAt.getMilliseconds() );

          const lastEvent = sortedEvents.at(-1)
          if(lastEvent?.kind !== kind || events.filter((event) => event.kind === kind).length > 1) return false;
          if(!beforeDate) return true;

          return lastEvent.occurredAt <= beforeDate
        })
      .map(({ siret }) => siret);
```